[LV] Vectorize selecting last IV of min/max element.

fhahn opened the pull request:
Add support for vectorizing loops that select the index of the minimum or
maximum element. The patch implements vectorizing those patterns by
combining Min/Max and FindFirstIV reductions.

It extends matching Min/Max reductions to allow in-loop users that are
FindLastIV reductions. It records a flag indicating that the Min/Max
reduction is used by another reduction.

When creating reduction recipes, we process any reduction that has other
reduction users. The reduction using the min/max reduction needs adjusting
to compute the correct result:
 1. We need to find the first IV for which the condition based on the
    min/max reduction is true,
 2. Compare the partial min/max reduction result to its final value and,
 3. Select the lanes of the partial FindLastIV reductions which correspond
    to the lanes matching the min/max reduction result.


Depends on https://github.com/llvm/llvm-project/pull/140451

Mel-Chen reviewed (commented):
I think it's a good idea that this patch avoids introducing a new RecurKind specifically for min/max with index.
However, I believe it would be more appropriate not to reuse AddReductionVars when idiom detection. The scalar phi for min/max doesn't really align with the current definition of a reduction. A reduction phi shouldn't be used by operations inside the loop that are not part of the recurrence chain, and it also needs to have users outside the loop.
So I think it should be detected independently.
Of course, a better approach would be to first identify the recurrence chain, and then determine whether it's a reduction.

Here's the source code we use for the detection part: https://github.com/llvm/llvm-project/pull/141467
As for the IR generation, I'd like to keep it internal for a while longer to test things before releasing it.

fhahn committed:
[LV] Vectorize select min/max index.

Add support for vectorizing loops that select the index of the minimum or
maximum element. The patch implements vectorizing those patterns by
combining Min/Max and FindFirstIV reductions.

It extends matching Min/Max reductions to allow in-loop users that are
FindLastIV reductions. It records a flag indicating that the Min/Max
reduction is used by another reduction.

When creating reduction recipes, we process any reduction that has other
reduction users. The reduction using the min/max reduction needs adjusting
to compute the correct result:
 1. We need to find the first IV for which the condition based on the
    min/max reduction is true,
 2. Compare the partial min/max reduction result to its final value and,
 3. Select the lanes of the partial FindLastIV reductions which correspond
    to the lanes matching the min/max reduction result.

fhahn commented:
> > So I think it should be detected independently. Of course, a better approach would be to first identify the recurrence chain, and then determine whether it's a reduction.
> 
> Yep that's a good point! I was initially planning to split off detection of the pattern separately as follow-up, but it seems better to do it to start with. I started sharing a few refactoring patches to make it simpler to create reduction phi recipes in VPlan (final one in the series is #142322) and will update the PR here in the next few days.

Updated the latest version to perform the check for the pattern in VPlan.

The main idea is that we allow unclassified phis in legal and hand off the VPlan to check if we can the remaining VPWidenPHIRecipes to min/max reductions used by FindLast/FirstIV. The current patch just supports the FindLastIV case, to not make it depend on https://github.com/llvm/llvm-project/pull/140451.

It now depends on https://github.com/llvm/llvm-project/pull/142322 to make it easy to create VPReductionPHIRecipes from VPlan w/o RecurrenceDescriptor. WDYT?

Mel-Chen commented:
> Updated the latest version to perform the check for the pattern in VPlan.
> 
> The main idea is that we allow unclassified phis in legal and hand off the VPlan to check if we can the remaining VPWidenPHIRecipes to min/max reductions used by FindLast/FirstIV. The current patch just supports the FindLastIV case, to not make it depend on #140451.
> 
> It now depends on #142322 to make it easy to create VPReductionPHIRecipes from VPlan w/o RecurrenceDescriptor. WDYT?

#142322 is good and will definitely be useful in the future. However, I’m a bit confused about VPlanTransforms::legalizeUnclassifiedPhis. Do we have any plans to move legality checks into the VPlan transformation phase? What are the benefits of doing this? 
I’m concerned that identifying min/max recurrences so late might limit future support for in-loop min/max reductions. Also, when min/max recurrences have LoopExitInstr, would this approach prevent us from using tail folding for vectorization?

Separately, I submitted #141467 and #142335, but they haven’t received any review yet. I’d like to understand if there are any major design issues with my patches that are preventing them from being accepted.

fhahn reviewed (commented):
All dependencies have now landed. In terms of reviews, would probably be best to first work through https://github.com/llvm/llvm-project/pull/146711, which uses a similar idea to vectrorize FMax w/o fast-math flags, but is slightly less complex over all I think

fhahn committed:
!fixup address review comments, thanks

fhahn commented:
ping :)

Mel-Chen commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9292,6 +9307,7 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
       PhiR->setOperand(0, StartV);
     }
   }
+
```
Remove the blank

fhahn replied:
Done thanks

Mel-Chen reviewed (commented):
I've been a bit busy lately, but I'll go through it thoroughly before this Friday.

fhahn committed:
!fixup remove stray new line

Mel-Chen commented:
I took the time to look at more examples. From what I’ve seen so far, deferring the decision to use the first or last index to VPlan seems beneficial. However, deferring the identification of min/max recurrence phis doesn’t appear to offer any clear advantage.

Would it be possible to still identify min/max recurrences during the legality stage, but postpone validating the interaction between the recurrence phi and FindIV until VPlan construction? 

My reasoning is:
1. Supporting min/max recurrences with loop-exiting users in scalable vectors is important to us. There’s a large-trip-count case in the SPEC benchmark suite that falls into this category. Early detection of min/max recurrence could help LoopVectorizationLegality::canFoldTailByMasking() behave more effectively. Or enable tail folding with live-outs before this patch lands.

2. The current approach builds VPlans even for phis that are clearly not recurrences and not related to min/max at all. But why build a VPlan for a loop that’s already known to be unvectorizable? 

Also, I noticed that by deferring recurrence detection to VPlan, we now need to check more types of recipes. This appears to introduce additional complexity.

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index

fhahn committed:
!fixup fix build after merge

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index-tmp

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index-tmp

fhahn committed:
!fixup detect multi-use min/max recurrences.

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index

fhahn committed:
!fixup add additional uses.

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index

fhahn reviewed (commented):
ping

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index

fhahn commented:
(supporting all min/max reduction cases as the previous implementation requires a bit of logic  than expected unfortunately, but it doesn't seem too bad)

fhahn committed:
!fixup replace MultiUse kinds with boolean flag

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index

fhahn committed:
Merge branch 'main' into lv-find-min-max-index

fhahn reviewed (commented):
Ping

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index

fhahn reviewed (commented):
ping for any additional comments

Mel-Chen commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -214,6 +214,40 @@ static bool checkOrderedReduction(RecurKind Kind, Instruction *ExactFPMathInst,
 }
 
+static std::optional<RecurrenceDescriptor>
+getMultiUseMinMax(PHINode *Phi, RecurKind Kind, Loop *TheLoop) {
```
Could not we modify the existing function for detecting min/max reductions to achieve this?

fhahn replied:
Probably yes, but currently this extension would be a bit messy I think, as the current implementation of `AddReductionVar` combines checking for all reduction kinds together, and adding another state to check that only applies to a subset seems to make it even more complicated generally.

I plan to unify it more, if/when https://github.com/llvm/llvm-project/pull/163460 lands.

Mel-Chen reviewed (commented):
I’d like to know if this patch depends on #163460?

fhahn reviewed (commented):
> I’d like to know if this patch depends on #163460?

Currently it does not. If/when #163460 land, we can easily merge them as part of  #163460 I think

ayalz commented on llvm/lib/Transforms/Utils/LoopUnroll.cpp:
```diff
@@ -1251,7 +1251,9 @@ llvm::canParallelizeReductionWhenUnrolling(PHINode &Phi, Loop *L,
-                                            /*AC=*/nullptr, /*DT=*/nullptr, SE))
+                                            /*AC=*/nullptr, /*DT=*/nullptr,
+                                            SE) ||
+      RdxDesc.isPhiMultiUse())
```
```suggestion
                                            /*AC=*/nullptr, /*DT=*/nullptr, SE))
    return std::nullopt;
  if (RdxDesc.isPhiMultiUse())
    return std::nullopt;
```
checking isPhiMultiUse() after isReductionPHI() returned true and initialized RdxDesc?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -965,3 +965,101 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+    RecurKind RdxKind = MinMaxPhiR->getRecurrenceKind();
+    // TODO: check for multi-uses in VPlan directly.
+    if (!RecurrenceDescriptor::isIntMinMaxRecurrenceKind(RdxKind) ||
+        !MinMaxPhiR->isPhiMultiUse())
```
Would checking the latter suffice, if it implies the former.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -965,3 +965,101 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+    if (!Sel ||
+        !match(Sel,
+               m_Select(m_Specific(Cmp), m_VPValue(IVOp), m_VPValue(FindIV))) ||
+        Sel->getNumUsers() != 2 || !isa<VPWidenIntOrFpInductionRecipe>(IVOp))
```
Could IVOp be scalar/derived?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -965,3 +965,101 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+        Sel->getNumUsers() != 2 || !isa<VPWidenIntOrFpInductionRecipe>(IVOp))
+      return false;
+    auto *FindIVPhiR = dyn_cast<VPReductionPHIRecipe>(FindIV);
+    if (!FindIVPhiR || !RecurrenceDescriptor::isFindLastIVRecurrenceKind(
```
If FindIVPhiR has a FindLastIVRecurrenceKind, does that imply IVOp must be a VPWidenIntOrFpInductionRecipe?

Mel-Chen commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7966,6 +7978,9 @@ void VPRecipeBuilder::collectScaledReductions(VFRange &Range) {
   for (const auto &[Phi, RdxDesc] : Legal->getReductionVars()) {
+    if (RecurrenceDescriptor::isMinMaxRecurrenceKind(
+            RdxDesc.getRecurrenceKind()))
+      continue;
```
Would the min/max recurrences originally be collected for partial reduction?

fhahn commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -225,6 +259,10 @@ bool RecurrenceDescriptor::AddReductionVar(
   if (Phi->getParent() != TheLoop->getHeader())
     return false;
 
+  if (auto RD = getMultiUseMinMax(Phi, Kind, TheLoop)) {
```
Yep updated the arguments, return value and naming, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index

fhahn committed:
!fixup address comments, thanks

ayalz commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -214,6 +214,51 @@ static bool checkOrderedReduction(RecurKind Kind, Instruction *ExactFPMathInst,
+  if (!Latch)
+    return false;
+  assert(Phi->getNumIncomingValues() == 2 &&
+         "phi must have exactly 2 incoming values");
```
nit: can slightly simplify
```suggestion
  assert(Phi->getNumIncomingValues() == 2 && "phi must have 2 incoming values");
```

fhahn replied:
done thanks

ayalz commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -214,6 +214,51 @@ static bool checkOrderedReduction(RecurKind Kind, Instruction *ExactFPMathInst,
+         "phi must have exactly 2 incoming values");
+  Value *Inc = Phi->getIncomingValueForBlock(Latch);
+  if (Phi->hasOneUse() ||
+      !RecurrenceDescriptor::isIntMinMaxRecurrenceKind(Kind))
```
nit: can slightly simplify by sinking to default case below
```suggestion
  if (Phi->hasOneUse())
```

fhahn replied:
Done thanks, but I retained the `isIntMinMaxRecurrenceKind` for now, to catch missing cases below

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8162,13 +8174,19 @@ VPRecipeBase *VPRecipeBuilder::tryToCreateWidenRecipe(VPSingleDefRecipe *R,
-    } else {
+          CM.useOrderedReductions(RdxDesc), ScaleFactor,
+          RdxDesc.isPhiMultiUse());
+    } else if (Legal->isFixedOrderRecurrence(Phi)) {
```
Ah yes, remove the if again, as this was still left over

fhahn commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -214,6 +214,51 @@ static bool checkOrderedReduction(RecurKind Kind, Instruction *ExactFPMathInst,
+    return false;
+
+  Value *A, *B;
+  bool Matched = [&]() {
```
done thanks

fhahn committed:
!fixup address comments, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7996,8 +8002,9 @@ void VPRecipeBuilder::collectScaledReductions(VFRange &Range) {
   for (const auto &[Phi, RdxDesc] : Legal->getReductionVars()) {
-    getScaledReductions(Phi, RdxDesc.getLoopExitInstr(), Range,
-                        PartialReductionChains);
+    if (Instruction *RdxExitInstr = RdxDesc.getLoopExitInstr()) {
```
Yep

fhahn commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -214,6 +214,51 @@ static bool checkOrderedReduction(RecurKind Kind, Instruction *ExactFPMathInst,
 }
 
+/// Returns true if \p Phi is a min/max reduction matching \p Kind where \p Phi
+/// is used in the loop outside the reduction chain. This is common for
```
Should be done now, thanks

ayalz commented on llvm/test/Transforms/LoopVectorize/AArch64/select-index.ll:
```diff
@@ -47,21 +47,70 @@ define i64 @test_vectorize_select_umin_last_idx(ptr %src, i64 %n) {
+; CHECK-NEXT:    [[IV:%.*]] = phi i64 [ 0, %[[VECTOR_PH]] ], [ [[INDEX_NEXT:%.*]], %[[VECTOR_BODY]] ]
+; CHECK-NEXT:    [[VEC_IND:%.*]] = phi <2 x i64> [ <i64 0, i64 1>, %[[VECTOR_PH]] ], [ [[VEC_IND_NEXT:%.*]], %[[VECTOR_BODY]] ]
+; CHECK-NEXT:    [[VEC_PHI:%.*]] = phi <2 x i64> [ splat (i64 -9223372036854775808), %[[VECTOR_PH]] ], [ [[TMP7:%.*]], %[[VECTOR_BODY]] ]
+; CHECK-NEXT:    [[VEC_PHI1:%.*]] = phi <2 x i64> [ splat (i64 -9223372036854775808), %[[VECTOR_PH]] ], [ [[TMP8:%.*]], %[[VECTOR_BODY]] ]
```
Better start at zero, see below.

ayalz commented on llvm/test/Transforms/LoopVectorize/AArch64/select-index.ll:
```diff
@@ -131,21 +180,70 @@ define i64 @test_vectorize_select_smin_last_idx(ptr %src, i64 %n) {
 ; CHECK:       [[LOOP]]:
-; CHECK-NEXT:    [[IV:%.*]] = phi i64 [ 0, %[[ENTRY]] ], [ [[IV_NEXT:%.*]], %[[LOOP]] ]
-; CHECK-NEXT:    [[MIN_IDX:%.*]] = phi i64 [ 0, %[[ENTRY]] ], [ [[MIN_IDX_NEXT:%.*]], %[[LOOP]] ]
-; CHECK-NEXT:    [[MIN_VAL:%.*]] = phi i64 [ 0, %[[ENTRY]] ], [ [[MIN_VAL_NEXT:%.*]], %[[LOOP]] ]
```
Independent- better initialize smin reduction with a negative number?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -992,3 +994,88 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+      return false;
+    auto MinMaxUsers = to_vector(MinMaxPhiR->users());
+    auto *Cmp = dyn_cast<VPRecipeWithIRFlags>(
+        MinMaxUsers[0] == MinMaxOp ? MinMaxUsers[1] : MinMaxUsers[0]);
```
Could this use some form of `findUser[W/ or W/O Opcode]()` as in `findComputeReductionResult()`?

ayalz replied:
Could `find_if(users, != MinMaxOp)` be used, as in `findReductionUser()`, for consistency?

fhahn replied:
updated to use new `findUser` with pattern matcher

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -818,13 +818,14 @@ void VPlanTransforms::addMinimumVectorEpilogueIterationCheck(
-static VPInstruction *
-findComputeReductionResult(VPReductionPHIRecipe *RedPhiR) {
-  auto It = find_if(RedPhiR->users(), [](VPUser *U) {
+/// If \p RedPhiR is used to compute are reduction result via a
```
Fixed, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -818,13 +818,14 @@ void VPlanTransforms::addMinimumVectorEpilogueIterationCheck(
-  auto It = find_if(RedPhiR->users(), [](VPUser *U) {
+/// If \p RedPhiR is used to compute are reduction result via a
+/// ComputeReductionResult, ComputeFindIVResult or ComputeAnyOfResult
+/// VPInstruction, return it. Otherwise return nullptr.
```
Yep

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -992,3 +994,88 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+        RecurrenceDescriptor::isIntMinMaxRecurrenceKind(RdxKind) &&
+        "only min/max recurrences support users outside the reduction chain");
+
+    // One user of MinMaxPhiR is MinMaxOp, the other users must be a compare
```
fixed, thanks

fhahn committed:
!fixup address comments, thanks

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/select-index.ll:
```diff
@@ -47,21 +47,70 @@ define i64 @test_vectorize_select_umin_last_idx(ptr %src, i64 %n) {
+; CHECK-NEXT:    [[TMP15:%.*]] = call i64 @llvm.vector.reduce.smax.v2i64(<2 x i64> [[RDX_MINMAX5]])
+; CHECK-NEXT:    [[RDX_SELECT_CMP:%.*]] = icmp ne i64 [[TMP15]], -9223372036854775808
+; CHECK-NEXT:    [[RDX_SELECT:%.*]] = select i1 [[RDX_SELECT_CMP]], i64 [[TMP15]], i64 0
+; CHECK-NEXT:    [[RDX_MINMAX6:%.*]] = call <2 x i64> @llvm.umin.v2i64(<2 x i64> [[TMP5]], <2 x i64> [[TMP6]])
```
Yep, no need to create a new ComputeReductionResult in the latest version, just updated to find + reuse the existing one, thanks

fhahn commented:
> > Currently it does not. If/when #163460 land, we can easily merge them as part of #163460 I think
> 
> Could we land #163460 and simplify this patch a bit?

I'd prefer to keep them separate for now, as more work would be needed to make sure the code in this PR can handle all cases supported by the logic in #163460  and it would probably be better to generalize once both landed

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -992,3 +993,87 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+    //  2. Compare the partial min/max reduction result to its final value and,
+    //  3. Select the lanes of the partial FindLastIV reductions which
+    //  correspond to the lanes matching the min/max reduction result.
+    auto *FindIVResult = cast<VPInstruction>(
```
Is the cast needed?

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -992,3 +993,87 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+    //  correspond to the lanes matching the min/max reduction result.
+    auto *FindIVResult = cast<VPInstruction>(
+        findReductionUser(FindIVPhiR, VPInstruction::ComputeFindIVResult));
+    auto *MinMaxResult = cast<VPInstruction>(
```
no, removed, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -818,13 +818,13 @@ void VPlanTransforms::addMinimumVectorEpilogueIterationCheck(
+  auto It = find_if(RedPhiR->users(), [Opcode](VPUser *U) {
     auto *VPI = dyn_cast<VPInstruction>(U);
-    return VPI && VPI->getOpcode() == VPInstruction::ComputeReductionResult;
+    return VPI && VPI->getOpcode() == Opcode;
```
Yes I added 2 variants: `findUser`, which takes a generic matcher (`findUser`,  and `findUserVPInstruction`).

I'll check for other potential users of the functions in a follow-up

fhahn committed:
!fixup address latest comments, thanks

fhahn committed:
!fixup add argmin/argmax tests with fmin/fmax.

ayalz commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -214,6 +214,51 @@ static bool checkOrderedReduction(RecurKind Kind, Instruction *ExactFPMathInst,
+  if (!IsMinMax)
+    return false;
+
+  if (A == B || (A != Phi && B != Phi))
```
Nit: the complement of either A == Phi xor B == Phi can be written as ((A == Phi) == (B == Phi)), but probably clearer as wrt.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -818,15 +818,19 @@ void VPlanTransforms::addMinimumVectorEpilogueIterationCheck(
+/// If \p V is used by a recipe matching pattern \p P, return it. Otherwise
+/// return nullptr;
+template <typename MatchT>
+static VPRecipeBase *findUser(VPValue *V, const MatchT &P) {
```
Perhaps `findUserOf()`, where "of" stands for both users of V and those of the given pattern P, corresponding to find_if.

fhahn replied:
updated, thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -992,3 +997,86 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+    VPValue *CmpOpB;
+    CmpInst::Predicate Pred;
+    auto *Cmp = dyn_cast_or_null<VPRecipeWithIRFlags>(
+        findUser(MinMaxPhiR, m_Cmp(m_VPValue(CmpOpA), m_VPValue(CmpOpB))));
```
```suggestion
        findUser(MinMaxPhiR, m_Cmp(m_VPValue(CmpOpA), m_VPValue())));
```
CmpOpB unused?

ayalz commented on llvm/test/Transforms/LoopVectorize/AArch64/select-index.ll:
```diff
@@ -93,7 +140,7 @@ define i64 @test_vectorize_select_smin_first_idx(ptr %src, i64 %n) {
 ; CHECK:       [[LOOP]]:
 ; CHECK-NEXT:    [[IV:%.*]] = phi i64 [ 0, %[[ENTRY]] ], [ [[IV_NEXT:%.*]], %[[LOOP]] ]
 ; CHECK-NEXT:    [[MIN_IDX:%.*]] = phi i64 [ 0, %[[ENTRY]] ], [ [[MIN_IDX_NEXT:%.*]], %[[LOOP]] ]
-; CHECK-NEXT:    [[MIN_VAL:%.*]] = phi i64 [ 0, %[[ENTRY]] ], [ [[MIN_VAL_NEXT:%.*]], %[[LOOP]] ]
```
Bumping initial value of reduction up from zero is mostly relevant for umin reduction.

ayalz commented on llvm/test/Transforms/LoopVectorize/select-fmax-last-index.ll:
```diff
@@ -0,0 +1,132 @@
+  %gep = getelementptr float, ptr %src, i64 %iv
+  %l = load float, ptr %gep
+  %cmp = fcmp uge float %max.val, %l
+  %max.cmp = fcmp ogt float %l, %max.val
```
Are these intentionally distinct compares.

ayalz commented on llvm/test/Transforms/LoopVectorize/select-fmax-last-index.ll:
```diff
@@ -0,0 +1,132 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --check-globals none --version 5
+; RUN: opt -passes=loop-vectorize -force-vector-width=4 -force-vector-interleave=1 -S %s | FileCheck %s
+
+define i64 @test_vectorize_select_fmaxnum_idx(ptr %src, i64 %n) {
```
May be good to continue and indicate if the desired `idx` is first or last, here and below.

fhahn replied:
Done, thanks

ayalz commented on llvm/test/Transforms/LoopVectorize/select-umin-last-index.ll:
```diff
@@ -305,21 +410,58 @@ define i64 @test_cmp_and_umin_use_different_values(ptr %src, i64 %x, i64 %n) {
+; CHECK-NEXT:    [[RES:%.*]] = phi i64 [ [[MIN_IDX_NEXT]], %[[LOOP]] ], [ [[RDX_SELECT]], %[[MIDDLE_BLOCK]] ]
 ; CHECK-NEXT:    ret i64 [[RES]]
 ;
 entry:
```
This records the last index when the umin reduction was greater or equal to %x (if any), excluding the last iteration, as it checks %min.val before it is updated by %1?

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7994,8 +7999,8 @@ void VPRecipeBuilder::collectScaledReductions(VFRange &Range) {
-    getScaledReductions(Phi, RdxDesc.getLoopExitInstr(), Range,
-                        ChainsByPhi[Phi]);
+    if (Instruction *RdxExitInstr = RdxDesc.getLoopExitInstr())
+      getScaledReductions(Phi, RdxExitInstr, Range, ChainsByPhi[Phi]);
```
done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -992,3 +997,86 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+    auto *MinMaxOp =
+        dyn_cast<VPRecipeWithIRFlags>(MinMaxPhiR->getBackedgeValue());
+    if (!MinMaxOp || MinMaxOp->getNumUsers() != 2)
+      return false;
```
added ,thanks

fhahn committed:
!fixup address latest comments, thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -965,3 +965,101 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+    // that's part of a FindLastIV chain.
+    auto *MinMaxOp =
+        dyn_cast<VPRecipeWithIRFlags>(MinMaxPhiR->getBackedgeValue());
+    if (!MinMaxOp || MinMaxOp->getNumUsers() != 2)
```
Better clarify what patterns are supported as accurately as possible at the outset.

Somewhat confusing - `isMinMaxReductionWithLoopUsersOutsideReductionChain()` checks that Phi has more than one user (in the loop) and Inc has one - here MinMaxOp (aka Inc?) has two?

fhahn replied:
Thanks, I added a comment after the initial early continue in the loop.

I added the assert (and more asserts later below checking the expected users)

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -992,3 +996,98 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+    if (MinMaxOpA != MinMaxPhiR)
+      std::swap(MinMaxOpA, MinMaxOpB);
+    if (MinMaxOpA != MinMaxPhiR)
+      return false;
```
Does `isMinMaxReductionWithLoopUsersOutsideReductionChain()` guarantee that either MinMaxOpA xor MinMaxOpB are MinMaxPhiR, i.e., should this be an assert.

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -992,3 +996,98 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+
+    // Normalize the predicate so MinMaxPhiR is on the right side.
+    if (CmpOpA == MinMaxPhiR)
+      Pred = CmpInst::getSwappedPredicate(Pred);
```
Not for the current version, but will be needed later to support more cases. I removed it for now, thanks

fhahn committed:
!fixup address latest comments, thanks

ayalz reviewed (approved):
LGTM, thanks for following up!

May be good to also have end-to-end test(s), to confirm vectorized result is the same as original; and subsequently check relative costs and overall performance benefits.

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index

fhahn commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -214,6 +214,52 @@ static bool checkOrderedReduction(RecurKind Kind, Instruction *ExactFPMathInst,
+    return false;
+
+  if (A == B || (A != Phi && B != Phi))
+    return false;
```
Ah yes, should be updated ,thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -1001,3 +1005,138 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+    if (MinMaxOpA != MinMaxPhiR)
+      std::swap(MinMaxOpA, MinMaxOpB);
+    assert(MinMaxOpA == MinMaxPhiR &&
+           "one of MinMaxOp's operands must be the phi");
```
Updated using `MinMaxOpValue`, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -1001,3 +1005,138 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+    assert(is_contained(MinMaxPhiR->users(), MinMaxResult) &&
+           "one user must be MinMaxResult");
+    assert(is_contained(MinMaxOp->users(), MinMaxPhiR) &&
+           "one user must be MinMaxPhiR");
```
dropped, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -1001,3 +1005,138 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+    assert(is_contained(MinMaxOp->users(), MinMaxPhiR) &&
+           "one user must be MinMaxPhiR");
+    assert(is_contained(MinMaxOp->users(), MinMaxResult) &&
+           "one user must be MinMaxResult");
```
done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -1001,3 +1005,138 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+    //
+    // into:
+    //
+    // vp<min.result> = compute-reduction-result ir<%min.val>, ir<%min.val.next>
```
yep, currently the ocpode includes a reference to the reduction phi recipe, to retrieve RdxDesc if needed

fhahn committed:
!fixup address comments, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index

fhahn committed:
!fixup

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index-tmp

fhahn committed:
Step

fhahn committed:
[VPlan] Use m_Intrinsic to match assumes/noalias_scope_decl (NFC).

fhahn committed:
[VPlan] Add matcher

fhahn committed:
Fix crash

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index-tmp

fhahn committed:
Merge remote-tracking branch 'origin/main' into lv-find-min-max-index-tmp

fhahn reviewed (commented):
> May be good to also have end-to-end test(s), to confirm vectorized result is the same as original; and subsequently check relative costs and overall performance benefits.

There was a set of tests in llvm-test-suite, but the coverage was relatively limited. Put up https://github.com/llvm/llvm-test-suite/pull/299 to generalize and extend coverage

mikaelholmen commented:
Hi @fhahn 
The following starts crashing with this patch:
```opt -passes=loop-vectorize bbi-113199.ll -o /dev/null```
Result:
```
opt: ../lib/Transforms/Vectorize/VPlanConstruction.cpp:1085: static bool llvm::VPlanTransforms::handleMultiUseReductions(VPlan &): Assertion `match(IVOp, m_TruncOrSelf(m_VPValue(IVOp))) && isa<VPWidenIntOrFpInductionRecipe>(IVOp) && "other select operand must be a (truncated) wide induction"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace and instructions to reproduce the bug.
Stack dump:
0.	Program arguments: build-all/bin/opt -passes=loop-vectorize bbi-113199.ll -o /dev/null
1.	Running pass "function(loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>)" on module "bbi-113199.ll"
2.	Running pass "loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>" on function "findLowestReg"
 #0 0x000056222b3095b6 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (build-all/bin/opt+0x4f5d5b6)
 #1 0x000056222b306b45 llvm::sys::RunSignalHandlers() (build-all/bin/opt+0x4f5ab45)
 #2 0x000056222b30a739 SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
 #3 0x00007fdf9f6b3990 __restore_rt (/lib64/libpthread.so.0+0x12990)
 #4 0x00007fdf9d05352f raise (/lib64/libc.so.6+0x4e52f)
 #5 0x00007fdf9d026e65 abort (/lib64/libc.so.6+0x21e65)
 #6 0x00007fdf9d026d39 _nl_load_domain.cold.0 (/lib64/libc.so.6+0x21d39)
 #7 0x00007fdf9d04be86 (/lib64/libc.so.6+0x46e86)
 #8 0x000056222ca2d59b llvm::VPlanTransforms::handleMultiUseReductions(llvm::VPlan&) (build-all/bin/opt+0x668159b)
 #9 0x000056222c949991 llvm::LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(std::unique_ptr<llvm::VPlan, std::default_delete<llvm::VPlan>>, llvm::VFRange&, llvm::LoopVersioning*) (build-all/bin/opt+0x659d991)
#10 0x000056222c93d014 llvm::LoopVectorizationPlanner::buildVPlansWithVPRecipes(llvm::ElementCount, llvm::ElementCount) (build-all/bin/opt+0x6591014)
#11 0x000056222c93c9c6 llvm::LoopVectorizationPlanner::plan(llvm::ElementCount, unsigned int) (build-all/bin/opt+0x65909c6)
#12 0x000056222c951588 llvm::LoopVectorizePass::processLoop(llvm::Loop*) (build-all/bin/opt+0x65a5588)
#13 0x000056222c95b56b llvm::LoopVectorizePass::runImpl(llvm::Function&) (build-all/bin/opt+0x65af56b)
#14 0x000056222c95be8a llvm::LoopVectorizePass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (build-all/bin/opt+0x65afe8a)
#15 0x000056222c81528d llvm::detail::PassModel<llvm::Function, llvm::LoopVectorizePass, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) PassBuilderPipelines.cpp:0:0
#16 0x000056222b52d045 llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (build-all/bin/opt+0x5181045)
#17 0x000056222c81293d llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) PassBuilderPipelines.cpp:0:0
#18 0x000056222b53195e llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x518595e)
#19 0x000056222c7a17dd llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) NewPMDriver.cpp:0:0
#20 0x000056222b52bd05 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x517fd05)
#21 0x000056222c799d04 llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::RTLIB::RuntimeLibcallsInfo&, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool, bool) (build-all/bin/opt+0x63edd04)
#22 0x000056222b2aad48 optMain (build-all/bin/opt+0x4efed48)
#23 0x00007fdf9d03f7e5 __libc_start_main (/lib64/libc.so.6+0x3a7e5)
#24 0x000056222b2a86ee _start (build-all/bin/opt+0x4efc6ee)
Abort (core dumped)
```
[bbi-113199.ll.gz](https://github.com/user-attachments/files/23834098/bbi-113199.ll.gz)

fhahn commented:
@mikaelholmen thanks for the report, should be fixed in cd3192a2c9c422f41d517428afef0a2232b9db8f by turing the assert into an early exit, with a TODO to handle selecting IV increments.
